src/Streamly/Internal/LZ4.hs

```diff
@@ -204,46 +253,74 @@ debug m = Stream.fromStreamD (debugD (Stream.toStreamD m))
               -- The size is safe to downcast
-              let srcLen1 = (fromIntegral :: Int -> CInt) srcLen
-              maxCLen <- c_compressBound srcLen1
-              when (maxCLen <= 0)
+              let uncompLenC = intToCInt uncompLen
+              maxCompLenC <- c_compressBound uncompLenC
+              let maxCompLen = cIntToInt maxCompLenC
+              when (maxCompLenC <= 0)
                 $ error $ "compressChunk: compressed length <= 0."
-                    ++ " maxCLen: " ++ show maxCLen
-                    ++ " source array length: " ++ show srcLen1
+                    ++ " maxCompLenC: " ++ show maxCompLenC
+                    ++ " uncompLenC: " ++ show uncompLenC
               -- allocate compressed block with 8 byte header.  First 4
               -- bytes of the header store the length of the uncompressed
               -- data and the next 4 bytes store the length of the
               -- compressed data.
               (MArray.Array fptr dstBegin dstMax) <-
-                  MArray.newArray (fromIntegral maxCLen + 8)
-              let hdrSrcLen :: Ptr Int32 = castPtr dstBegin
-                  hdrCompLen :: Ptr Int32 = dstBegin `plusPtr` 4
-                  compData = dstBegin `plusPtr` 8
-              compLen <-
+                  MArray.newArray (maxCompLen + metaLen)
+              let hdrCompLen = dstBegin `plusPtr` compLenOff
+                  compData = dstBegin `plusPtr` compDataOff
+              compLenC <-
                   c_compressFastContinue
-                      ctx src compData srcLen1 maxCLen (fromIntegral i)
-              when (compLen <= 0)
+                      ctx src compData uncompLenC maxCompLenC speedC
+              when (compLenC <= 0)
                 $ error $ "compressChunk: c_compressFastContinue failed. "
-                    ++ "Source array length: " ++ show srcLen1
-                    ++ "Return value: " ++ show compLen
-              poke hdrSrcLen (coerce srcLen1)
-              poke hdrCompLen (coerce compLen)
-              let dstEnd = dstBegin `plusPtr`
-                            ((fromIntegral :: CInt -> Int) compLen + 8)
+                    ++ "uncompLenC: " ++ show uncompLenC
+                    ++ "compLenC: " ++ show compLenC
+              encodeUncompLen (castPtr dstBegin) (cIntToI32 uncompLenC)
+              poke hdrCompLen (cIntToI32 compLenC)
+              let compLen = cIntToInt compLenC
+                  dstEnd = dstBegin `plusPtr` (compLen + metaLen)
                   compArr = MArray.Array fptr dstEnd dstMax
               Array.unsafeFreeze <$> MArray.shrinkToFit compArr
 
+    where
+
+    maxUncompLen = min maxUncompressedSize (cIntToInt lz4_MAX_INPUT_SIZE)
+
+    metaLen =
+        if encodeUncompressedSize
+        then 8
+        else 4
+
+    compLenOff =
+        if encodeUncompressedSize
+        then 4
+        else 0
+
+    compDataOff =
+        if encodeUncompressedSize
+        then 8
+        else 4
```

Comment:
These three functions can be made top level and reused in both compress/decompress.